### DISCUSSION TITLE
bitmap fonts

### DISCUSSION THREAD

**Discussion by oskude** [slint-ui/slint#2381] [Q&A]
can we use bitmap fonts in (embedded rust) slint?

or other way to get crisp fonts on smaller sizes?

ps. fwiw, i was able to use pcf fonts with micropython, nanogui and https://github.com/peterhinch/micropython-font-to-py - but now i found slint and can't go back :heart:

**Comment by tronical** [comment]
Right now bitmap fonts are unfortunately not supported out of the box. That's ironic because the glyphs pre-rendered at compile-time are bitmaps (well, alpha-maps :-). But I agree, we should support them.

If it's okay with you, I'll convert this discussion into an issue, because this is something that should be implemented in Slint.

I see different ways of doing this:

1. Supporting bitmaps embedded in TrueType fonts: When embedded via sbix or clbc+clbt tables, we could extract them via [ttf-parser](https://docs.rs/ttf-parser/0.18.1/ttf_parser/struct.Face.html?search=bitmap#method.glyph_raster_image). This has the advantage that the font is indexed by fontdb and the selection/font discovery mechanism remains unified.
2. We could support bdf fonts using the https://docs.rs/bdf-reader/latest/bdf_reader/index.html crate. This is straight-forward to use, but we need to introduce a separate discovery mechanism for the .bdf files.

I'm unsure about pcf - that would probably require writing a PCF parser. Or does anyone know of a crate out there that supports it?


@oskude How would you feel about converting your PCF font to a TrueType font with embedded bitmaps? Do you know if there are any good tools? (I'm asking because that could also be documented in Slint)

**Comment by oskude** [comment]
> How would you feel about converting your PCF font to a TrueType font with embedded bitmaps? Do you know if there are any good tools?

I'm fine with converting to TrueType bitmap font. And especially as Slint already uses TTF, sounds like the best choice!

For font tools i've only used https://fontforge.org

To test i opened a PCF font from https://terminus-font.sourceforge.net/ and created a TTF from it with:

![Screenshot from 2023-03-23 19-20-32](https://user-images.githubusercontent.com/157669/227311917-50d7db0e-8e9f-4f40-ba97-f79f880ecd6d.png)

> or with script `makettf.pe`:
> ```fontforge
> #!/usr/bin/fontforge
> Open($1)
> Generate($1:r + ".ttf")
> ```
> and running: `./makettf.pe ter-x12n.pcf`
>> note: apparently the hashbang script name has to end with ".pe", or fontforge gives a syntax error.

And this Slint code compiles:
```slint
import "./TerminusMedium.ttf";

export component MainWindow inherits Window {
    default-font-family: "Terminus";
    default-font-size: 12px;
    ...
```

There is no text shown in app, i assume as expected, cause the TTF has only bitmap fonts. But otherwise everything else is shown and works, cool!

If useful, here's the TTF file it generated:
[terminus-12-ttf.zip](https://github.com/slint-ui/slint/files/11054736/terminus-12-ttf.zip)

PS. I'm a font (and rust and embedded) newbie, but i'm happy to test or help anything :sweat_smile:
